cinder-csi-plugin leaves kubelet in retry-unmount-loops under some circumstances

<!-- This form is for bug reports and feature requests! -->

**Is this a BUG REPORT or FEATURE REQUEST?**:

/kind bug

**The binaries affected**:

- [ ] openstack-cloud-controller-manager
- [x] cinder-csi-plugin
- [ ] k8s-keystone-auth
- [ ] client-keystone-auth
- [ ] octavia-ingress-controller
- [ ] manila-csi-plugin
- [ ] manila-provisioner
- [ ] magnum-auto-healer
- [ ] barbican-kms-plugin

**What happened**:

Using `volumeMode: Block` results in dangling `volumesInUse` on the node object after deleting the Pod. This causes `kubelet` to try unounting the volume over and over again, without success.

```
kubectl get node k8s-node-1 -oyaml
...
status:
  volumesInUse:
  - kubernetes.io/csi/cinder.csi.openstack.org^2cf05bab-9a12-4af5-b660-1965a56aed81
```

`kubelet` logs show the attempt to unmount the volume. It gets a "NotFound" error even though the still exists in openstack:
```
Jan 07 11:27:14 k8s-node-1 kubelet[622]: E0107 11:27:14.130514     622 nestedpendingoperations.go:270] Operation for "\"kubernetes.io/csi/cinder.csi.openstack.org^2cf05bab-9a12-4af5-b660-1965a56aed81\"" failed. No retries permitted until 2020-01-07 11:27:46.130454676 +0000 UTC m=+457.111885005 (durationBeforeRetry 32s). Error: "UnmapDevice.TearDownDevice failed for volume \"pvc-27a5b2b0-166c-4140-9541-b443315309ab\" (UniqueName: \"kubernetes.io/csi/cinder.csi.openstack.org^2cf05bab-9a12-4af5-b660-1965a56aed81\") on node \"k8s-node-1\" : kubernetes.io/csi: blockMapper.unstageVolumeForBlock failed: rpc error: code = NotFound desc = Volume not mounted"
```

The volume effectively is unmounted on that node:

```
root@k8s-node-1:~# mount | grep csi
root@k8s-node-1:~#
```

This behavior also causes failures in the e2e test of CSIMigration, resulting in orphaned OpenStack Volumes (volume still exists in OpenStack, but there is no representation in k8s any more).

(See https://github.com/theopenlab/openlab-zuul-jobs/blob/master/playbooks/cloud-provider-openstack-multinode-csi-migration-test/run.yaml for the calling command)

```
Jan 07 09:47:05 control-plane run.sh[6621]: Summarizing 5 Failures:
Jan 07 09:47:05 control-plane run.sh[6621]: [Fail] [sig-storage] In-tree Volumes [Driver: cinder] [Testpattern: Dynamic PV (block volmode)] multiVolume [Slow] [It] should access to two volumes with the same volume mode and retain data across pod recreation on different node
Jan 07 09:47:05 control-plane run.sh[6621]: /workspace/anago-v1.17.0-rc.2.10+70132b0f130acc/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/storage/testsuites/multivolume.go:360
Jan 07 09:47:05 control-plane run.sh[6621]: [Fail] [sig-storage] In-tree Volumes [Driver: cinder] [Testpattern: Dynamic PV (block volmode)] multiVolume [Slow] [It] should access to two volumes with different volume mode and retain data across pod recreation on the same node
Jan 07 09:47:05 control-plane run.sh[6621]: /workspace/anago-v1.17.0-rc.2.10+70132b0f130acc/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/storage/testsuites/multivolume.go:120
Jan 07 09:47:05 control-plane run.sh[6621]: [Fail] [sig-storage] In-tree Volumes [Driver: cinder] [Testpattern: Dynamic PV (block volmode)] volumes [It] should store data
Jan 07 09:47:05 control-plane run.sh[6621]: /workspace/anago-v1.17.0-rc.2.10+70132b0f130acc/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/storage/testsuites/volumes.go:146
Jan 07 09:47:05 control-plane run.sh[6621]: [Fail] [sig-storage] In-tree Volumes [Driver: cinder] [Testpattern: Dynamic PV (block volmode)] multiVolume [Slow] [It] should access to two volumes with different volume mode and retain data across pod recreation on different node
Jan 07 09:47:05 control-plane run.sh[6621]: /workspace/anago-v1.17.0-rc.2.10+70132b0f130acc/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/storage/testsuites/multivolume.go:360
Jan 07 09:47:05 control-plane run.sh[6621]: [Fail] [sig-storage] In-tree Volumes [Driver: cinder] [Testpattern: Dynamic PV (block volmode)] multiVolume [Slow] [It] should access to two volumes with the same volume mode and retain data across pod recreation on the same node
Jan 07 09:47:05 control-plane run.sh[6621]: /workspace/anago-v1.17.0-rc.2.10+70132b0f130acc/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/storage/testsuites/multivolume.go:120
Jan 07 09:47:05 control-plane run.sh[6621]: Ran 46 of 4814 Specs in 3951.775 seconds
Jan 07 09:47:05 control-plane run.sh[6621]: FAIL! -- 41 Passed | 5 Failed | 0 Pending | 4768 Skipped
Jan 07 09:47:05 control-plane run.sh[6621]: --- FAIL: TestE2E (3951.92s)
```

**What you expected to happen**:

When deleting a Pod with a volumeDevice attached ...
* the volume must be unmounted
* `kubelet` doesn't end up in a retry loop
* all traces (`volumeattachment` and `status.volumesInUse`) are cleared
* e2e test of CSIMigration succeed

**How to reproduce it**:

Deploy cinder-csi-plugin and a csi storageclass

```
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: csi-sc-cinderplugin
  provisioner: cinder.csi.openstack.org
```

Create a PVC usind `volumeMode: block`

```
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: mycsiclaim
  namespace: manual-testing
spec:
  accessModes:
    - ReadWriteOnce
  volumeMode: Block
  resources:
    requests:
      storage: 1Gi
  storageClassName: csi-sc-cinderplugin
```

Create a Pod using the PVC

```
apiVersion: v1
kind: Pod
metadata:
  name: task-pv-pod-csi
  namespace: manual-testing
spec:
  volumes:
    - name: task-pv-storage
      persistentVolumeClaim:
        claimName: mycsiclaim
  containers:
    - name: task-pv-container
      image: nginx
      ports:
        - containerPort: 80
          name: "http-server"
      volumeDevices:
        - devicePath: /dev/foo
          name: task-pv-storage
```

Wait for the pod to come up.

Verify the settings on the node object
```
  volumesAttached:
  - devicePath: ""
    name: kubernetes.io/csi/cinder.csi.openstack.org^2cf05bab-9a12-4af5-b660-1965a56aed81
  volumesInUse:
  - kubernetes.io/csi/cinder.csi.openstack.org^2cf05bab-9a12-4af5-b660-1965a56aed81
```

When the pod is up and running, the volume is attached, delete the pod

```
kubectl delete pod task-pv-pod-csi
```

From this on, you will see the mentioned log messages in `kubelet`. The corresponding `volumeattachment` remains in place for quite a while, even though the volume was unmounted successfully.
```
kubectl get volumeattachment
NAME                                                                   ATTACHER                   PV                                         NODE         ATTACHED   AGE
csi-b9f748a84231b1e130ad4ddba226c21584c49a1eacbe1d891c7f00b0e8b5a3cd   cinder.csi.openstack.org   pvc-27a5b2b0-166c-4140-9541-b443315309ab   k8s-node-1   true       6m55s
```

Grab a coffee and enjoy it for around 5-10 minutes. After that time, the `volumeattachment` is gone, but the `volumesInUse`-field is still set, and `kubelet` retries to unmount over and over again.
While `volumesAttached` is empty/removed.

```
  volumesInUse:
  - kubernetes.io/csi/cinder.csi.openstack.org^2cf05bab-9a12-4af5-b660-1965a56aed81
```

**Anything else we need to know?**:

According to the CSI spec, the nodeplugin should return `NotFound` in case the `volume_id` doesn't exists. Which here means, the Volume in OpenStack, not the local mount.

This results in always returning `NotFound` on a second call to NodeUnstageVolume. What potentially can happen when an corresponding object changes. So in theory this is not limited to `volumeMode: Block` but also for `Filesystem`.

See:
* https://github.com/container-storage-interface/spec/blob/master/spec.md#nodeunstagevolume-errors
* https://github.com/container-storage-interface/spec/blob/master/spec.md#nodeunpublishvolume-errors

Note: restarting kubelet, also resolves the dangling `volumesInUse`.

**Environment**:
- openstack-cloud-controller-manager version: latest (0c132d1db92852337665fe309759b713209ff32d)
- OpenStack version: Rocky
- Kubernetes version: v1.17.0

--- a/pkg/csi/cinder/nodeserver.go
+++ b/pkg/csi/cinder/nodeserver.go
@@ -284,10 +284,13 @@ func (ns *nodeServer) NodeUnpublishVolume(ctx context.Context, req *csi.NodeUnpu
 	m := ns.Mount
 	notMnt, err := m.IsLikelyNotMountPointDetach(targetPath)
 	if err != nil && !mount.IsCorruptedMnt(err) {
+		klog.V(4).Infof("NodeUnpublishVolume: unable to unmount volume: %v", err)
 		return nil, status.Error(codes.Internal, err.Error())
 	}
 	if notMnt && !mount.IsCorruptedMnt(err) {
-		return nil, status.Error(codes.NotFound, "Volume not mounted")
+		// the volume is not mounted at all. There is no need for retrying to unmount.
+		klog.V(4).Infoln("NodeUnpublishVolume: skipping... not mounted any more")
+		return &csi.NodeUnpublishVolumeResponse{}, nil
 	}
 
 	err = m.UnmountPath(targetPath)
@@ -416,6 +419,7 @@ func (ns *nodeServer) NodeUnstageVolume(ctx context.Context, req *csi.NodeUnstag
 	_, err := ns.Cloud.GetVolume(volumeID)
 	if err != nil {
 		if cpoerrors.IsNotFound(err) {
+			klog.V(4).Infof("NodeUnstageVolume: Unable to find volume: %v", err)
 			return nil, status.Error(codes.NotFound, "Volume not found")
 		}
 		return nil, status.Error(codes.Internal, fmt.Sprintf("GetVolume failed with error %v", err))
@@ -425,10 +429,13 @@ func (ns *nodeServer) NodeUnstageVolume(ctx context.Context, req *csi.NodeUnstag
 
 	notMnt, err := m.IsLikelyNotMountPointDetach(stagingTargetPath)
 	if err != nil && !mount.IsCorruptedMnt(err) {
+		klog.V(4).Infof("NodeUnstageVolume: unable to unmount volume: %v", err)
 		return nil, status.Error(codes.Internal, err.Error())
 	}
 	if notMnt && !mount.IsCorruptedMnt(err) {
-		return nil, status.Error(codes.NotFound, "Volume not mounted")
+		// the volume is not mounted at all. There is no need for retrying to unmount.
+		klog.V(4).Infoln("NodeUnstageVolume: skipping... not mounted any more")
+		return &csi.NodeUnstageVolumeResponse{}, nil
 	}
 
 	err = m.UnmountPath(stagingTargetPath)

